### PR TITLE
BTC Better fee estimation in UnspentSelector

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

In Bitcoin plan, first a number of UTXOs are selected, then the fee is estimated.
The second step uses proper SegWit estimation (by previewing encoded TX and computing vSize).
The first step, however, used the non-segwit estimation (143 x noin+34 x noout + 10).
As a result, in certain cases plan() could not find a solution, when in fact there was one.
Now in Bitcoin case, the rough estimation used in the first step is changed to the more segwit-correct 101.25 x noin + 31 x noout + 10.

The tradeoff here:
- if higher fee is used (pre-change), it is possible with certain inputset/amount combinations that we fail to create transaction, when in fact a valid segwit-fee transaction exists
- if lower fee is used (post-change), and transaction is NOT segwit, yet we use segwit estimation in selection, finalk fee will be higher, resulting change will be lower, or actual fee will be lower, resulting in a slow tx.

## Testing instructions

Unit tests.

## Types of changes

* Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

**Comment by catenocrypt** [comment]
Should come after #1248 .

**Review by vikmeup — APPROVED** [review]

**Comment by vikmeup** [comment]
This algorithm only applied for SegWit supported chains? As some blockchain like doge coin does not support SegWit

**Comment by catenocrypt** [comment]
> This algorithm only applied for SegWit supported chains? As some blockchain like doge coin does not support SegWit

This change applies now to BTC and LTC only.

File: src/Bitcoin/FeeCalculator.cpp [hunk 1]
```diff
@@ -62,6 +72,10 @@ FeeCalculator& getFeeCalculator(TWCoinType coinType) {
     case TWCoinTypeDecred:
         return decredFeeCalculator;
 
+    case TWCoinTypeBitcoin:
+    case TWCoinTypeLitecoin:
```
Comment (hewigovens) [inline src/Bitcoin/FeeCalculator.cpp:76]: and: viacoin, digibyte, bitcoin gold

File: src/Bitcoin/FeeCalculator.cpp [hunk 2]
```diff
@@ -21,6 +21,15 @@ int64_t DefaultFeeCalculator::calculateSingleInput(int64_t byteFee) const {
     return int64_t(148) * byteFee;
 }
 
+int64_t SegwitFeeCalculator::calculate(int64_t inputs, int64_t outputs, int64_t byteFee) const {
+    const auto txsize = int64_t(std::ceil(101.25 * inputs + 31.0 * outputs + 10.0));
+    return txsize * byteFee;
+}
+
+int64_t SegwitFeeCalculator::calculateSingleInput(int64_t byteFee) const {
+    return int64_t(std::ceil(101.25)) * byteFee;
```
Comment (hewigovens) [inline src/Bitcoin/FeeCalculator.cpp:30]: just use a const `102` to remove all `std::ceil` calls

Comment (catenocrypt) [reply src/Bitcoin/FeeCalculator.cpp:30]: Modern compilers are smart enough to replace this to 102 at compile time, but I changed as you suggest to be sure :)

**Review by hewigovens — APPROVED** [review]

**Comment by catenocrypt** [comment]
wc 2.5.5

**Pull request merged**